tracing-core/src/collect.rs

```diff
@@ -277,6 +282,17 @@ pub trait Collect: 'static {
     /// follow from _b_), it may silently do nothing.
     fn record_follows_from(&self, span: &span::Id, follows: &span::Id);
 
+    /// Determine if an [`Event`] should be recorded.
+    ///
+    /// By default, this returns `true` and collectors can filter events in
+    /// [`event`][Self::event] without any penalty. However, when `event` is
+    /// more complicated, this can be used to determine if `event` should be
+    /// called at all, separating out the decision from the processing.
+    fn event_enabled(&self, event: &Event<'_>) -> bool {
+        let _ = event;
+        true
+    }
```

Comment:
Okay, another note: `tracing` and `tracing-core` will _never_ call this; it's currently used *only* by `tracing-subscriber`. Maybe we should change `Dispatch` to call this prior to `event`...but maybe we should just document it? I don't _love_ adding code in `core` that's specifically for `tracing-subscriber` to consume, but :woman_shrugging:

Reply:
maybe we could rename the `Collect` method to something that frames it as a "hint", like `would_enable_event` or something? or maybe that just makes it more confusing. I dunno...

Reply:
/shrug

Also it's quite probable that there are forwarding Collect impls which I haven't forwarded `event_enabled` for. However, `Dispatch` should probably still call `event_enabled`, so `Collect` impls which implement `event_enabled` do the right thing.

Reply:
It's worth noting that *requiring* checking `event_enabled` is a *forwards* compatibility concern: previous code won't call `event_enabled` and will just call `event`, so perhaps it *should* be worded as a hint... but then tracing-subscriber using it to control whether things higher on the stack see the event seems kind of sketchy as well.